src/Features/CSharp/Portable/CodeFixes/AddExplicitCast/AddExplicitCastCodeFixProvider.cs

```diff
@@ -0,0 +1,372 @@
+            var diagnostic = context.Diagnostics.First();
+
+            var root = await document.GetRequiredSyntaxRootAsync(context.CancellationToken).ConfigureAwait(false);
+            var syntaxFacts = document.GetRequiredLanguageService<ISyntaxFactsService>();
+            var semanticModel = await document.GetRequiredSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+
+            var targetNode = root.FindNode(diagnostic.Location.SourceSpan, getInnermostNodeForTie: true)
+                .GetAncestorsOrThis<ExpressionSyntax>().FirstOrDefault();
+            if (targetNode != null)
+            {
+                var hasSolution = GetTypeInfo(semanticModel, root, targetNode, cancellationToken, out var nodeType, out var potentialConversionTypes);
+                if (hasSolution && potentialConversionTypes.Length == 1)
+                {
+                    context.RegisterCodeFix(new MyCodeAction(
+                        CSharpFeaturesResources.Add_explicit_cast,
+                        c => FixAsync(context.Document, context.Diagnostics.First(), c)),
+                        context.Diagnostics);
+                }
+                else if (hasSolution && potentialConversionTypes.Length > 1)
+                {
+                    var actions = ArrayBuilder<CodeAction>.GetInstance();
+
+                    // MaximumConversionOptions: we show at most [MaximumConversionOptions] options for this code fixer
+                    for (var i = 0; i < Math.Min(MaximumConversionOptions, potentialConversionTypes.Length); i++)
+                    {
+                        var convType = potentialConversionTypes[i];
+                        actions.Add(new MyCodeAction(string.Format(CSharpFeaturesResources.Convert_type_to_0, convType.ToDisplayString()),
+                            c => ApplyFixAsync(context.Document, root, targetNode, convType)));
+                    }
+
+                    // If the number of potential conversion types is larger than options we could show, report telemetry
+                    if (potentialConversionTypes.Length > MaximumConversionOptions)
+                    {
+                        Logger.Log(FunctionId.CodeFixes_AddExplicitCast,
+                            KeyValueLogMessage.Create(m =>
+                            {
+                                m["NumberOfCandidates"] = potentialConversionTypes.Length;
+                            }));
+                    }
+
+                    context.RegisterCodeFix(new CodeAction.CodeActionWithNestedActions(
+                        CSharpFeaturesResources.Add_explicit_cast,
+                        actions.ToImmutableAndFree(), false),
+                        context.Diagnostics);
+                }
+            }
+        }
+
+        private static async Task<Document> ApplyFixAsync(Document document, SyntaxNode currentRoot, ExpressionSyntax targetNode, ITypeSymbol conversionType)
```

Comment:
this method doesn't seem to be async at all. Remove `async` modifier, and replace last like with `return Task.FromResult(document.With...));`